README.md
PyOMAPIc
========

Python bindings to [ISC DHCP OMAPI](http://ipamworldwide.com/dhcp-api.html) library.

On amd64 (x86_64) platforms libomapi, libds and libdst have to be compiled
with the -fPIC option.

Based on Python OMAPI (DHCP) Interface by Perry Lorier <perry@coders.net> and
Matt Brown <matt@mattb.net.nz>: http://source.meta.net.nz/svn/dhcparpd/trunk/pyomapi/ (not avaliable now)

Many thanks to Ignace Mouzannar (-ghantoos-) for his article on creating a deb from python distutils.
http://ghantoos.org/2008/10/19/creating-a-deb-package-from-a-python-setuppy/


debian/changelog
python-omapic (2.0) UNRELEASED; urgency=low

  * New functions in the PyOMAPIc module

 -- root <root@lenny-amd64-builder>  Tue, 20 Oct 2009 13:21:19 +0400

python-omapic (1.99) UNRELEASED; urgency=low

  * Initial release. (Closes: #XXXXXX)

 -- root <root@lenny-amd64-builder>  Tue, 20 Oct 2009 12:33:00 +0400


test.py
#!/usr/bin/env python
from omapi import Omapi
a = Omapi('192.168.1.1',7911,'NET_UPDATER', 'HHJ7346IOAAAAAAAAAAAAAA==')
#print a.lookup_ip('0:16:76:c9:14:9b')
#print a.lookup_mac('172.22.12.30')
#print a.lookup_host_ip('00:16:76:c9:14:9b')
print a.lookup_host_ip('0:15:C5:68:5E:75')
#print a.lookup_host_mac('172.22.12.4')
#print a.add_host('172.22.12.6','11:22:33:44:55:66')
#print a.add_host('172.22.12.6','11:22:33:44:55:66')
#print a.lookup_host_ip('11:22:33:44:55:66')
#print a.lookup_host_mac('172.22.12.6')
#print a.lookup_lease_mac('172.22.18.19')
#print a.del_host(None, '172.22.12.6')
#print a.del_host('11:22:33:44:55:66',None)
#print a.add_host('172.22.12.6','11:22:33:44:55:66')
#print a.del_host('11:22:33:44:55:66','172.22.12.6')
#print a.add_host('172.22.12.6','11:22:33:44:55:66')
#print a.del_host('11:22:33:44:55:66','')
#print a.del_host('11:22:33:44:55:66','')
#print a.lookup_host_mac('172.22.12.2')
#


trunk/PyOMAPIc.py
#!/usr/bin/env python

from omapi import Omapi
import re

ipre = re.compile('^([0-9]{1,3}\.){3}[0-9]{1,3}$')
macre = re.compile('^([0-9a-fA-F]{1,2}\:){5}[0-9a-fA-F]{1,2}$')


class PyOMAPIc(Omapi):
	"""
	Python wrapper for C-based functions for OMAPI (DHCP)
	"""

	def _normalizeMac(self,mac):
		if not mac: return ''
		if not re.match(macre,mac): return mac
		try: return ":".join(map(lambda x:"%02x"%x,map(lambda x: int(x,16),mac.split(":"))))
		except: return mac
	
	def getLeaseMac(self,ip):
		"""
		Returns MAC-address for lease with selected IP-address
		"""
		return self._normalizeMac(self.lookup_lease_mac(ip))
	
	def getHostMac(self,ip):
		"""
		Returns MAC-address for host with selected IP-address
		"""
		return self._normalizeMac(self.lookup_host_mac(ip))
	
	def getMac(self,ip):
		"""
		Returns MAC-address for active lease, or for host if lease not found.
		"""
		result = self.getLeaseMac(ip)
		if result: return result
		return self.getHostMac(ip)
	
	def getHostsMacs(self,ips):
		"""
		Returns hosts list [(ip,mac),...] for IPs list
		"""
		result = []
		for ip in ips: result.append((ip,self.getHostMac(ip)))
		return result
	
	def getLeasesMacs(self,ips):
		"""
		Returns leases list [(ip,mac),...] for IPs list
		"""
		result = []
		for ip in ips: result.append((ip,self.getLeaseMac(ip)))
		return result
	
	def getMacs(self,ips):
		"""
		Returns hosts or leases list [(ip,mac),...] for IPs list
		"""
		result = []
		for ip in ips: result.append((ip,self.getMac(ip)))
		return result

	def addHost(self,ip,mac):
		"""
		Creates new host with specified MAC and IP addresses
		"""
		bip = self.lookup_host_ip(mac)
		if bip and bip != ip: raise ValueError("MAC address %s is already bound to %s"%(mac,bip))
		bmac = self.getHostMac(ip)
		if bmac and bmac != self._normalizeMac(mac): raise ValueError("IP address %s is already associated with the MAC %s"%(ip, bmac))
		if not bip == ip and not bmac == mac:
			self.add_host(ip, mac)
		return 'OK'
	
	def delHost(self,ip,mac=None):
		"""
		Deletes a host by specified IP/MAC
		"""
		self.del_host(mac,ip)
		return 'OK'
	


base64.c
/*    Base64 encode/decode
 *    Copyright (C) 2005  Perry Lorier
 *
 *    This program is free software; you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation; either version 2 of the License, or
 *    (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program; if not, write to the Free Software
 *    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * Compile with -DTEST for a standalone test program
 */

#include <stdlib.h>
#include <inttypes.h>
#include <string.h>

typedef unsigned char uchar_t;
uchar_t *base64_encode(const uchar_t *buf, size_t len)
{
	const unsigned char *str="ABCDEFGHJIKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
	uchar_t *ret=malloc((len*4+2)/3+1);
	const uchar_t *cp=buf;
	uchar_t *out=ret;
	while(len>2) {
		uint32_t val=(cp[0]<<16)|(cp[1]<<8)|(cp[2]);
		cp+=3;
		len-=3;

		/* AAAAAAxx xxxxxxxx xxxxxxxx */	
		*(out++)=str[(val>>18)&0x3F];

		/* xxxxxxAA AAAAxxxx xxxxxxxx */
		*(out++)=str[(val>>12)&0x3F];

		/* xxxxxxxx xxxxAAAA AAxxxxxx */
		*(out++)=str[(val>> 6)&0x3F];

		/* xxxxxxxx xxxxxxxx xxAAAAAA */
		*(out++)=str[(val    )&0x3F];
	}

	if (len == 2) {
		uint32_t val=(cp[0]<<16)|(cp[1]<<8);
		cp+=2;
		len-=2;

		/* AAAAAAxx xxxxxxxx xxxxxxxx */	
		*(out++)=str[(val>>18)&0x3F];

		/* xxxxxxAA AAAAxxxx xxxxxxxx */
		*(out++)=str[(val>>12)&0x3F];

		/* xxxxxxxx xxxxAAAA AAxxxxxx */
		*(out++)=str[(val>> 6)&0x3F];

		/* xxxxxxxx xxxxxxxx xxAAAAAA */
		*(out++)='=';
	}
	else if (len == 1) {
		uint32_t val=(cp[0]<<16);
		cp+=2;
		len-=2;

		/* AAAAAAxx xxxxxxxx xxxxxxxx */	
		*(out++)=str[(val>>18)&0x3F];

		/* xxxxxxAA AAAAxxxx xxxxxxxx */
		*(out++)=str[(val>>12)&0x3F];

		/* xxxxxxxx xxxxAAAA AAxxxxxx */
		*(out++)='=';

		/* xxxxxxxx xxxxxxxx xxAAAAAA */
		*(out++)='=';
	}

	*(out++)='\0'; /* remember to NUL terminate the string */

	return ret;
}

const uchar_t base64_decode_matrix[] = { 
/*	 0  1  2  3  4  5  6  7  8  9  A  B  C  D  E  F          */
	99,99,99,99,99,99,99,99,99,99,99,99,99,99,99,99, /* 0x00 */
	99,99,99,99,99,99,99,99,99,99,99,62,63,99,99,99, /* 0x10 */
	99,99,99,99,99,99,99,99,99,99,99,62,63,99,99,99, /* 0x20 */
	52,53,54,55,56,57,58,59,60,61,99,99,99,99,99,99, /* 0x30 */
	99, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9,10,11,12,13,14, /* 0x40 */
	15,16,17,18,19,20,21,22,23,24,25,99,99,99,99,99, /* 0x50 */
	99,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40, /* 0x60 */
	41,42,43,44,45,46,47,48,49,50,51,99,99,99,99,99, /* 0x70 */
	99,99,99,99,99,99,99,99,99,99,99,99,99,99,99,99, /* 0x80 */
	99,99,99,99,99,99,99,99,99,99,99,99,99,99,99,99, /* 0x90 */
	99,99,99,99,99,99,99,99,99,99,99,99,99,99,99,99, /* 0xA0 */
	99,99,99,99,99,99,99,99,99,99,99,99,99,99,99,99, /* 0xB0 */
	99,99,99,99,99,99,99,99,99,99,99,99,99,99,99,99, /* 0xC0 */
	99,99,99,99,99,99,99,99,99,99,99,99,99,99,99,99, /* 0xD0 */
	99,99,99,99,99,99,99,99,99,99,99,99,99,99,99,99, /* 0xE0 */
	99,99,99,99,99,99,99,99,99,99,99,99,99,99,99,99  /* 0xF0 */
	};

void base64_decode(const uchar_t *in, uchar_t **out, int *outlen)
{
	int inlen = strlen(in);
	int i;
	int j=0;

	if (inlen < 2) {
		*out=malloc(0);
		*outlen=0;
		return;
	}

	if (in[inlen-2]=='=') {
		*outlen=inlen*3/4-2;
	}
	else if (in[inlen-1]=='=') {
		*outlen=inlen*3/4-1;
	}
	else {
		*outlen=inlen*3/4;
	}
	*out=malloc(*outlen);
	for(i=0;i<inlen;) {
		int val=0;
		/* Skip invalid charactors (eg whitespace) */
		while (i<inlen && base64_decode_matrix[in[i]]==99) {
			++i;
		}
		if (i>=inlen)
			break;
		val=base64_decode_matrix[in[i++]]<<18;
		/* Skip invalid charactors (eg whitespace) */
		while (i<inlen && base64_decode_matrix[in[i]]==99) {
			++i;
		}
		if (i>=inlen)
			break;

		val|=base64_decode_matrix[in[i++]]<<12;
		(*out)[j++]=val>>16;
		
		/* Skip invalid charactors (eg whitespace) */
		while (i<inlen && base64_decode_matrix[in[i]]==99) {
			++i;
		}
		if (i>=inlen)
			break;

		val|=base64_decode_matrix[in[i++]]<<6;
		(*out)[j++]=val>>8;

		/* Skip invalid charactors (eg whitespace) */
		while (i<inlen && base64_decode_matrix[in[i]]==99) {
			++i;
		}
		if (i>=inlen)
			break;

		val|=base64_decode_matrix[in[i++]];
		(*out)[j++]=val;

	}
}

#ifdef TEST
#include <stdio.h>

int main(int argc, char *argv[])
{
	uchar_t *buf;
	int len;
	if (argc<2) {
		fprintf(stderr,"Usage: base64 string\n");
		return 1;
	}
	printf("%s\n",base64_encode(argv[1],strlen(argv[1])));
	base64_decode(base64_encode(argv[1],strlen(argv[1])),&buf,&len);
	printf("%s\n",buf);

	return 0;
}
#endif
